### PR TITLE
[sharedtree] remove dead code

### PR DESCRIPTION
**Pull request by honzasterba**

### CODE REVIEW

**Comment by michalkurka** [comment]
FIY: I have a pending branch that is affecting by these changes, can you wait with merge after I merge my PR?

File: h2o-algos/src/main/java/hex/tree/BuildHistogram.java [hunk 1]
```diff
@@ -170,102 +203,115 @@ public void setupLocal() {
...
+    private void map(int id, Chunk [] chks) {
+      final C4VolatileChunk nids = (C4VolatileChunk) chks[_nidIdx];
+      // Pass 1: Score a prior partially-built tree model, and make new Node
+      // assignments to every row.  This involves pulling out the current
+      // assigned DecidedNode, "scoring" the row against that Node's decision
+      // criteria, and assigning the row to a new child UndecidedNode (and
+      // giving it an improved prediction).
+      int [] nnids;
+      if( _leaf > 0)            // Prior pass exists?
+        nnids = score_decide(chks,nids.getValues());
```
Comment (maurever) [inline h2o-algos/src/main/java/hex/tree/BuildHistogram.java:227]:
```suggestion
        nnids = scoreDecide(chks,nids.getValues());
```

**Comment by honzasterba** [comment]
will wait, please link your PR here

File: h2o-algos/src/main/java/hex/tree/BuildHistogram.java [hunk 2]
```diff
@@ -170,102 +203,115 @@ public void setupLocal() {
...
+        if (i >= 0)
+          nh[i + 1]++;
+      // Rollup the histogram of rows-per-NID in this chunk
+      for (int i = 0; i <_numLeafs; i++) nh[i + 1] += nh[i];
+      // Splat the rows into NID-groups
+      int rows[] = (_rss[id] = new int[nnids.length]);
+      for (int row = 0; row < nnids.length; row++)
+        if (nnids[row] >= 0)
+          rows[nh[nnids[row]]++] = row;
+
```
Comment (maurever) [inline h2o-algos/src/main/java/hex/tree/BuildHistogram.java:250]:
```suggestion

```

File: h2o-algos/src/main/java/hex/tree/BuildHistogram.java [hunk 3]
```diff
@@ -170,102 +203,115 @@ public void setupLocal() {
...
+      // This is how it works:
+      //    1) Outer MRTask walks down it's tree, forking tasks with exponentially decreasing number of columns until reaching its left most leaf for columns 0.
+      //       At this point, the local fjq for this thread has a task for processing half of columns at the bottom, followed by task for 1/4 of columns and so on.
+      //       Other threads start stealing work from the bottom.
+      //    2) forks the leaf task and (because its polling from the top) executes the LocalMr for the column 0.
+      // This way we should have columns as equally distributed as possible without resorting to shared priority queue
+      new LocalMR(new MrFun() {
+        @Override
+        protected void map(int c) {
+          c = active_cols == null?c:active_cols[c];
```
Comment (maurever) [inline h2o-algos/src/main/java/hex/tree/BuildHistogram.java:310]:
```suggestion
          c = active_cols == null ? c : active_cols[c];
```

File: h2o-algos/src/main/java/hex/tree/BuildHistogram.java [hunk 4]
```diff
@@ -170,102 +203,115 @@ public void setupLocal() {
...
+          new LocalMR(new ComputeHistogramTask(_hcs.length == 0?new DHistogram[0]:_hcs[c],c,largestChunkSz,new AtomicInteger()),numWrks + (c < rem?1:0), BuildHistogram.this).fork();
```
Comment (maurever) [inline h2o-algos/src/main/java/hex/tree/BuildHistogram.java:311]: Could you split this command into more small ones? Or is it memory critical to not save some parts to a variable?  It is very difficult to read.

File: h2o-algos/src/main/java/hex/tree/BuildHistogram.java [hunk 5]
```diff
@@ -57,25 +57,76 @@
  *    exp(nthreads-pre-column) = max(1,H2O.NUMCPUS - num_cols)
  *
  */
-public class ScoreBuildHistogram2 extends ScoreBuildHistogram {
-  transient int []   _cids;
+public class BuildHistogram extends MRTask<BuildHistogram> {
```
Comment (maurever) [inline h2o-algos/src/main/java/hex/tree/BuildHistogram.java:60]: Why did you decide to remove 'Score' from the name? It still scores histogram, or not?

File: h2o-algos/src/main/java/hex/tree/BuildHistogram.java [hunk 6]
```diff
@@ -170,102 +203,115 @@ public void setupLocal() {
...
-        for (int i = 0; i <_numLeafs; i++) nh[i + 1] += nh[i];
-        // Splat the rows into NID-groups
-        int rows[] = (_rss[id] = new int[nnids.length]);
-        for (int row = 0; row < nnids.length; row++)
-          if (nnids[row] >= 0)
-            rows[nh[nnids[row]]++] = row;
+    new LocalMR(new PhaseOne(), new PhaseTwo(this, largestChunkSz)).fork();
+  }
+  
+  private class PhaseOne extends MrFun<PhaseOne> {
```
Comment (maurever) [inline h2o-algos/src/main/java/hex/tree/BuildHistogram.java:214]: It is a really nice change to have each pass as separate classes. However, I would prefer the classes have names that describe what they do. If not, add a brief docstring, please. 

I suggested something like ScoreAndSortPhase and BuildPhase or something like that.

**Review by maurever — CHANGES_REQUESTED** [review]
Thanks for this big move to have more readable code. I have just a few suggestions. 

All tests passed? It is a critical part of code for all tree algos, so you should also check benchmarks.

File: h2o-algos/src/main/java/hex/tree/BuildHistogram.java (hunk 5 above)
Comment (honzasterba) [reply h2o-algos/src/main/java/hex/tree/BuildHistogram.java:60]: yes it does but it is not `scoring` in the sense that is used elsewehere, I think the `BuildHistogram` name is just fine

File: h2o-algos/src/main/java/hex/tree/BuildHistogram.java (hunk 4 above)
Comment (honzasterba) [reply h2o-algos/src/main/java/hex/tree/BuildHistogram.java:310]: will do

File: h2o-algos/src/main/java/hex/tree/BuildHistogram.java (hunk 5 above)
Comment (michalkurka) [reply h2o-algos/src/main/java/hex/tree/BuildHistogram.java:60]: yes, but this class is full of "scoring" terminology in javadoc - imho it correctly reflects the 2 phase approach - it does more than just to build the histograms, it is not an idempotent action

Comment (honzasterba) [reply h2o-algos/src/main/java/hex/tree/BuildHistogram.java:60]: adding back Score to the name

**Comment by honzasterba** [comment]
will be split into parts and re-done

**Pull request closed without merging**